### PR TITLE
prometheus proxy managed under tigera operator

### PR DESCRIPTION
**Pull request by freecaykes**
## For PR author

- [x] Tests for change.
- [x] If changing pkg/apis/, run `make gen-files`
- [x] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [x] Milestone set according to targeted release.
- [x] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: config/enterprise_versions.yml
Comment (freecaykes) [inline config/enterprise_versions.yml:86]: - [ ] cleanup before merge

File: pkg/components/enterprise.go
Comment (freecaykes) [inline pkg/components/enterprise.go:136]: - [ ] cleanup before merge

File: pkg/components/references.go
Comment (freecaykes) [inline pkg/components/references.go:53]: - [ ] cleanup before merge

File: pkg/render/prometheus_service.go
Comment (doublek) [inline pkg/render/prometheus_service.go:154]: nit: some typos in this comment. WDYTA "Set to host networked if EKS with Calico CNI because the EKS managed Kubernetes APIserver cannot reach the pod network in this configuration. This is because Calico cannot manage the managed control plane nodes' network" or something like that.

File: pkg/render/prometheus_service.go
Comment (doublek) [inline pkg/render/prometheus_service.go:163]: @mgleung do you think we need some PSP's for this?

**Review by doublek — COMMENTED** [review]
Some initial comments. 

- We should wait to get sign off from @tmjd and/or @caseydavenport regarding how to configure the port of the proxy. 
- Would like to get @mgleung's opinion to see if we need to put some PSP's.

File: pkg/render/monitor.go
Comment (freecaykes) [reply pkg/render/monitor.go]: settled with `prometheus-http-api`

File: api/v1/monitor_types.go
Comment (freecaykes) [inline api/v1/monitor_types.go:32]: wondering if this is the right pattern to configure  the port since it will only configure an otherwise hidden  configuration (ie. serves connections through the calico-node-pometheus Service target port to the listen port of the Prometheus Proxy )  

another option is the configmap

 @tmjd or @caseydavenport  wdyt

Comment (caseydavenport) [reply api/v1/monitor_types.go:32]: > In the scenario that the cluster is hosted on EKS and using Calico as it's CNI

I'd like to understand this more. Is this the only reason why we need the port to be configurable by the user?

The reason we make some things host-networked in this mode is because the EKS control-plane nodes attempt to send connections to those pods. Do the EKS control plane nodes initiate connections to the prometheus pod?

File: pkg/render/monitor.go
Comment (caseydavenport) [inline pkg/render/monitor.go:170]: nit: acronyms should be same-case in golang. So, `prometheusHTTPAPIService` is the right casing here.

File: pkg/render/prometheus_service.go
Comment (caseydavenport) [inline pkg/render/prometheus_service.go:47]: calling this PrometheusService is a tad confusing to me, since a `Service` is a specific kind of thing in Kubernetes, and this renderer creates more than just a `Service` (for example, it also has a `Deployment`).

File: pkg/render/prometheus_service.go
Comment (caseydavenport) [reply pkg/render/prometheus_service.go:162]: For what it is worth, PSPs are deprecated and are being removed. https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/

The replacement isn't available yet, so maybe still need them here.

File: api/v1/monitor_types.go
Comment (freecaykes) [reply api/v1/monitor_types.go:32]: >  Is this the only reason why we need the port to be configurable by the user?

yeah basically that is the major reason since prometheus proxy will be hostnetworked in that cluster configuration only, for the scenario that customer's might also have conflict with another hostnorked pod bind to this port

 I guess with exposing this configuration it can be configured at any time at the user's discretion  -> not sure if it's better to ignore this config if not hostnetwked

> Do the EKS control plane nodes initiate connections to the prometheus pod?
manager and voltron using the EKS managed kube apiserver as a proxy is the one connecting to the prometheus proxy pod

File: pkg/render/prometheus_service.go
Comment (freecaykes) [reply pkg/render/prometheus_service.go:47]: yeah sorry naming this was hard, using `Service` because the proxy might potentially evolve to something  closer to a full fledged api service, maybe i'll change to naming it to `tigera-prometheus-api`

File: api/v1/monitor_types.go
Comment (caseydavenport) [reply api/v1/monitor_types.go:32]: ick. horrible.

Comment (caseydavenport) [reply api/v1/monitor_types.go:32]: I don't see why we'd use a configmap for this - we have a CR available and it's valid user config. 

What's mostly sad is that we need this config option at all for a pretty niche scenario.

File: pkg/render/monitor.go
Comment (freecaykes) [reply pkg/render/monitor.go]: https://github.com/tigera/operator/pull/1379/commits/4ec097420c2e67725222b5db380aecff96ca1727

File: pkg/render/prometheus_service.go
Comment (freecaykes) [reply pkg/render/prometheus_service.go]: https://github.com/tigera/operator/pull/1379/commits/4ec097420c2e67725222b5db380aecff96ca1727

File: pkg/render/prometheus_api.go
Comment (tmjd) [inline pkg/render/prometheus_api.go:158]: Please leave Service out of the name here. I'd be fine if this function was reduced to `proxyDeployment`, especially if the component has `tigeraPrometheus` in the name.

File: pkg/render/prometheus_api.go
Comment (tmjd) [inline pkg/render/prometheus_api.go:217]: similar comment here about dropping `tigeraPrometheus` from the function name.

**Review by tmjd — COMMENTED** [review]
I've reviewed as an operator maintainer (being concerned about structure and fitting with operator conventions). I'd ask that you have another review that reviews for meeting design and functionality goals.
One thing I would have thought I'd see is a change to the prometheus resources that were currently being deployed because i thought the previous calico node prometheus service would need to be changed as part of these changes.

File: pkg/controller/monitor/monitor_controller.go
Comment (freecaykes) [reply pkg/controller/monitor/monitor_controller.go]: https://github.com/tigera/operator/pull/1379/commits/0aecaf91feb429b1e0bf6e1e0d25be76e044121e

File: config/enterprise_versions.yml
Comment (freecaykes) [reply config/enterprise_versions.yml:89]: yeah think really screwed up with the name for this, since semaphore is already deploying the image as prometheus-service and the repo is of the same name,  but that's basically why, since it might become a full fledged service / api for prometheus with RBAC and more endpoints than a proxy

we should find a common name and change it everywhere 

not sure if you're ok with `tigera-prometheus-api` or agree on a different name @tmjd

File: pkg/render/prometheus_api.go
Comment (freecaykes) [reply pkg/render/prometheus_api.go]: https://github.com/tigera/operator/pull/1379/commits/0aecaf91feb429b1e0bf6e1e0d25be76e044121e

File: pkg/render/prometheus_api.go
Comment (freecaykes) [reply pkg/render/prometheus_api.go]: https://github.com/tigera/operator/pull/1379/commits/0aecaf91feb429b1e0bf6e1e0d25be76e044121e

File: pkg/render/prometheus_api.go
Comment (freecaykes) [reply pkg/render/prometheus_api.go]: https://github.com/tigera/operator/pull/1379/commits/0aecaf91feb429b1e0bf6e1e0d25be76e044121e

File: api/v1/monitor_types.go
Comment (tmjd) [reply api/v1/monitor_types.go]: I missed this conversation before. Was this before or after we had the call discussing the configuration of the port? @caseydavenport are you expecting this configuration to be in a ConfigMap or part of the Monitor resource?

File: pkg/render/prometheus_api.go
Comment (tmjd) [inline pkg/render/prometheus_api.go:100]: Sorry I was not clear about this but getting this configmap should still be done as part of the monitor reconcile. It was only the creation of the configmap should be done by the render if it did not already exist (if nil was passed in for the configmap to the render).

File: pkg/controller/monitor/monitor_controller.go
Comment (tmjd) [reply pkg/controller/monitor/monitor_controller.go]: I commented about this in the render already but the get above should still be done in the monitor_controller.

File: config/enterprise_versions.yml
Comment (tmjd) [reply config/enterprise_versions.yml:89]: This is fine since the image/repo/ci has already been created. Don't worry about changing it.

File: pkg/render/prometheus_api.go
Comment (tmjd) [reply pkg/render/prometheus_api.go]: This wasn't shortened any, WDYT about just changing it to deployment?

File: pkg/render/prometheus_api.go
Comment (tmjd) [reply pkg/render/prometheus_api.go]: Same here, I think you should shorten it to `podSecurityPolicy`

File: pkg/render/prometheus_api.go
Comment (freecaykes) [reply pkg/render/prometheus_api.go]: sorry misunderstood last time, I think like the naming of everything related to the componenet to just be its kind, same with PSP below

File: pkg/render/prometheus_api.go
Comment (freecaykes) [reply pkg/render/prometheus_api.go]: https://github.com/tigera/operator/pull/1379/commits/15a53feb00afb31e12a5c3523cc9ea60837b0c2e

File: pkg/render/prometheus_api.go
Comment (freecaykes) [reply pkg/render/prometheus_api.go:56]: reason is related back to this comment https://github.com/tigera/operator/pull/1379#discussion_r675930250 

goes back  to expanding its purpose to more than just a proxy, so chose API, do we want to call it as is  or have it recognized as proxy in operator ? wdyt @tmjd

File: pkg/render/prometheus_api.go
Comment (tmjd) [inline pkg/render/prometheus_api.go:113]: I think adding a PSP might need to be dependent on not being on openshift, at least from looking at other renders.
Have you tried this on an openshift install?

Comment (freecaykes) [reply pkg/render/prometheus_api.go]: thanks for catching that, it didnt end up deploying the images with PSPs on in openshift 

https://github.com/tigera/operator/pull/1379/commits/9f54439f513d948e141eaec7de4808449f8d3b53

File: pkg/render/prometheus_api_test.go
Comment (tmjd) [inline pkg/render/prometheus_api_test.go:297]: It doesn't look like this is being used, could you remove it?

**Review by tmjd — APPROVED** [review]
LGTM

**Review by doublek — APPROVED** [review]
LGTM

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

**Pull request merged**
